### ISSUE TITLE
spatial expression data and H&E images alignment

### ISSUE THREAD

**Issue by wwYinYin** [klarman-cell-observatory/HTAPP-Pipelines#3]
Excellent work. The article says “the spatial expression data and H&E images were aligned in a semi-manual process to honor their serial nature and allow efficient comparison as well as transfer of histopathological annotations from the H&E images.” Could you please share the alignment code? Or provide the aligned data? Thank you very much.

**Comment by jklughammer** [comment]
Hi, the anndata objects available on scp and cellxgene already contain the aligned coordinates/HE images. If you use the provided coordinates of data and HEs of the same sample to plot in the same space you will see that they are aligned.
Beyond that the code used to aid the alignment can be found here: https://github.com/klarman-cell-observatory/HTAPP-Pipelines/blob/54a13eb8ab29d1b56d97ba2cb4c6df9f7c9fc24d/HTAPP_MBC/03_spatial_analysis/spatial_utils.py#L152
And here you can follow the individual alignment process of each sample: HTAPP_MBC/03_spatial_analysis/00.1_processing

**Comment by peipp410** [comment]
Hello! I have another question regarding the feasibility of treating the registered data as a form of spatial multi-omics data for integrated analysis.

My understanding is that after successful registration, we have spatial coordinates (e.g., `x`, `y` in `adata.obs`) that are aligned across different modalities.  My specific question is:

Can we consider the registered data as a foundation for multi-omics analysis, where, for a given spot on the histology image and its corresponding coordinates (x, y), we can link data from different spatial platforms (e.g., Stereo-seq for gene expression and CODEX for protein expression)?

For example, if I have a specific spot on the registered HE image, could I perform the following workflow:

1.  Identify the corresponding spatial coordinates (x, y) in the AnnData object (`adata.obs`).
2.  Retrieve the gene expression data from Stereo-seq associated with those coordinates.
3.  Retrieve the protein expression data from CODEX associated with the same coordinates.
4.  Extract a small patch of the HE image centered around the spot of interest.
5.  Analyze the relationship between the histological morphology (from the image patch) and the gene/protein expression levels (from Stereo-seq and CODEX) for that specific location.

Thank you for your time and guidance!

**Comment by jklughammer** [comment]
Yes, technically what you suggest works since the different data types have been aligned. However, it is necessary to consider that the different data types were generated form different serial sections i.e. not the exact same section - therefor there may be biological differences between data types (data in the same spot most likely don't com from the same cells). You can estimate the amount of change in the z direction by comparing the two HE images available for each metastasis (or the replicates when available). The sectioning scheme always follows the same order (as depicted in Fig 1), in 2 consecutive sets (indicated in the data by _1 = first set, and _2 = second set).

**Comment by jklughammer** [comment]
Oh and we don't have Stereo-seq - just Slide-seq

**Comment by peipp410** [comment]
> Oh and we don't have Stereo-seq - just Slide-seq

Oh yes, I apologize for the confusion. Thank you for your response!

**Comment by peipp410** [comment]
Sorry, just one more question. I noticed that, after registering and projecting the CODEX data onto the corresponding H&E images, some points appear outside the tissue slice boundary? Thanks!

![Image](https://github.com/user-attachments/assets/ab113268-1c4e-4fa8-815b-115c39201c44)

**Comment by jklughammer** [comment]
Yes, since HE and data don't come from exactly the same sections these inaccuracies are unfortunately what we have to work with. Here, we only used linear transformations to register which cannot align warped structure. Better alignment may be achieved using methods like STalign which use landmark-based non-linear transformation. However, these come at the cost of also distorting internal structure which may distort cellular neighborhoods.

**Comment by peipp410** [comment]
I see. Thanks!
